### ISSUE TITLE
ShennonFano&Huffman

### ISSUE THREAD

**Issue by Alexandr611** [Alexandr611/university#1]
<!DOCTYPE html>
<html>
<head 	>
	<meta charset="utf-8">
	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

	
	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">
	<style type="text/css">
	@import url('https://fonts.googleapis.com/css?family=Lato');
		form{
			position: relative;
			top:125px;
			min-height: 350px;
			height: auto;
			width: 400px;
			border: 1px solid #4285F4;
		}
		ol{
			padding: 0;
		}
		li{	
			margin: 7px 0;
			background: #5CB85C;
			border-radius: 5px;
			list-style: none;
		}
		span{
			text-decoration: center;
			padding-left: 25px;
			font-size: 22px;
			color: white;
		}
		form input{
			padding-left: 25px;
			margin-bottom: 25px;
		}
		#Input_text{
			font-size: 18px;
			border:none;
			border-bottom: 1px solid #4285F4;
			height: 30px;
		}
		#Input_text:hover{
			border: 1px solid #4285F4;
		}
		#demo{
			color: black;
		}
		form{
			padding: 25px;
		}
		#btn{
			margin-right: 8px;
			color: #4285F4;
			padding: 6px 20px;
			background: white;
			border: 1px solid #4285F4;
			border-radius: 3px;
		}
		#btn:hover{
			color: white;
			background: #4285F4;
		}
		.col-md-3{
			height: 1500px;
			background: #4285F4;
		}
		.col-md-3 li{
			background: transparent;
		}
		.col-md-3 li span:hover{
			cursor: pointer;
		}
		.col-md-3 li span{
			font-family: 'Lato', sans-serif;
			color: white;
			font-size: 25px;
		}
		.container{
			position: relative;
			top: 150px;
		}
		.progress-bar{
			background: #5CB85C;
		}
		.download{
			position: relative;
			top: 150px;
		}
		.file_upload{
    display: block;
    position: relative;
    overflow: hidden;
    font-size: 1em;              /* example */
    height: 2em;                 /* example */
    line-height: 2em             /* the same as height */
}
.file_upload .button, .file_upload > mark{
    display: block;
    cursor: pointer              /* example */
}
.file_upload .button{
    float: right;
    box-sizing: border-box;
    -moz-box-sizing: border-box;
    width: 8em;                  /* example */
    height: 100%;
    text-align: center           /* example */
}
.file_upload > mark{
    background: transparent;     /* example */
    padding-left: 1em            /* example */
}
@media only screen and ( max-width: 500px ){  /* example */
    .file_upload > mark{
        display: none
    }
    .file_upload .button{
        width: 100%
    }
}
.file_upload input[type=file]{
    position: absolute;
    top: 0;
    opacity: 0
}

/* Making it beautiful */

.file_upload{
    font: 14px/1 "Helvetica Neue", "Helvetica", Helvetica,Arial, sans-serif; /* example */
    border: 1px solid #ccc;
    border-radius: 3px;
    box-shadow: 0 0 5px rgba(0,0,0,0.1);
    transition: box-shadow 0.1s linear
}
.file_upload.focus{
    box-shadow: 0 0 5px rgba(0,30,255,0.4)
}
.file_upload .button{
    background: #7300df;
    transition: background 0.2s;
    border: 1px solid rgba(0,0,0,0.1);
    border-color: rgba(0,0,0,0.1) rgba(0,0,0,0.1) rgba(0,0,0,0.25);
    border-radius: 2px;
    box-shadow: 0 1px 0 rgba(255, 255, 255, 0.2) inset, 0 1px 2px rgba(0, 0, 0, 0.05);
    color: #fff;
    text-shadow: #6200bd 0 -1px 0;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis
}
.file_upload:hover .button{
    background: #6200bd;
    text-shadow: #5d00b3 0 -1px 0
}
.file_upload:active .button{
    background: #5d00b3;
    box-shadow: 0 0 3px rgba(0,0,0,0.3) inset
}
	</style>
	<script type="text/javascript">
		var text;
	 	var openFile = function(event) {
     	var input = event.target;

     	var reader = new FileReader();
	     	reader.onload = function(){
	       	text = reader.result;
	       	console.log(reader.result);
	       	document.getElementById('result').innerHTML = text;
     	};
     	reader.readAsText(input.files[0]);
	   	};
	   	var assoc_arr = new Array();
		var char_temp = new symbolShennon('','',0);
		var string_arr = new Array();
		var temp = 0;
		var count = 0;
		var demo = "";
		var length;
		function symbolShennon (str, code, count) {
	    this.str = str;
	    this.code = code;
	    this.count = count;
		}
	   	var method;
		var assoc_arr = [];
		var obj = [];
		var temp = 0;
		var count = 0;
		var demo = "";
		function bulbSort(char,count1){
		for (var i = 1; i < count1; i++){ 
	 		for (var j = 0; j < count1 - i; j++){
    			if (char[j].count < char[j+1].count){
				 		char_temp = char[j+1];
				 		char[j+1] = char[j];
				 		char[j] = char_temp; 
				 	}
				}
		 	}
		}
		function consoleLog(char,count){
			var testerino= '';
			for(var i = 0 ;i < count; i++){
				console.log(char[i].str);
				console.log(char[i].count);
			}
		}
		function huffmanMethod(char,string_arr,temp){
			for(var i = 0 ; i < char[temp].str.length; i++){
				for(var j = 0 ; j < string_arr.length; j++){
					if( string_arr[j].str == char[temp].str[i]){
						string_arr[j].code = '0'+string_arr[j].code;
					}
				}
			}
			if(char[temp-1] != null){
				for(var i = 0 ; i < char[temp-1].str.length ; i++){
					for(var j = 0 ; j < string_arr.length; j++){
						if( string_arr[j].str == char[temp-1].str[i]){
							string_arr[j].code = '1'+string_arr[j].code;
						}
					}
				}
			char[temp-1].str +=	char[temp].str;
			char[temp-1].count += char[temp].count;
			}
			
			// for(var i = 0 ; i < char[temp].str.length ; i++){
			// 	for(var j = 0 ; j < string_arr.length;j++){
					
			// 	}
			// }
			// if(char[temp-1] != null){
			// }
			  // string_arr[].code += "1";
			  // string_arr[-1].code += "0";
			return char;
		}
		function code(str,code){
			this.str = str;
			this.code = code;
		}
		function symbol (str, count, code) {
	    this.str = str;
	    this.count = count;
	    this.code= code;
		}
		function Shannon_Fano(first,last){
	var i, temp;
    var p;
    var pfull;
    var phalf;
    if( first == last ){
        return ;
        alert("exit");
    }
    else if( last - first == 1 ){
        string_arr[first].code += '0';
        string_arr[last].code += '1';
    }
    else{
        pfull = 0;
        for(i = first ; i <= last; ++i ){
        pfull += string_arr[i].count;
        }
        p = 0;
        temp = -1;
        phalf = pfull * 0.5;
        for( i= first; i <= last; ++i ){
            p += string_arr[i].count;
            if(p <= phalf) {
            	string_arr[i].code += '0';
            }
            else{
                string_arr[i].code += '1';
                if( temp < 0 ) temp = i;
            }
        }
            if( temp < 0 ) temp = first+1;

            Shannon_Fano( first, temp-1 );
    		Shannon_Fano( temp, last );
    		}
	}
		function decoder(string_arr,text){
			for(var i = 0;i<string_arr.length;i++){
				for(var j = 0;j<text.length;j++){
					if(string_arr[i].str == text[j]){
							text[j] = string_arr[i].code;
					}
				}
			}
			var demo = '';
			for(var i = 0;i< text.length;i++){
				demo += text[i];
			}
			document.getElementById('result').innerHTML = demo;
		}
		function encoder(string_arr,text){
			for(var i = 0;i<string_arr.length;i++){
				for(var j = 0;j<text.length;j++){
					if(string_arr[i].code == text[j]){
						text[j] = string_arr[i].str;
					}
				}
			}
			var demo = '';
			for(var i = 0;i<text.length;i++){
				demo += text[i];
			}
			document.getElementById('result').innerHTML = demo;
		}
		function activateHuffman(){
			method = true;
			alert("method Huffman activated!");
		}
		function activateShennon_Fano(){
			method = false;
			alert("method Shennon activated!");
		}
		function checkForm(form){
			if(method){
				char = new Array();
				var container = '';
				char_tmp = new Array();
				//text = form.inputName.value;
				text = text.split('');
				length = text.length;
				char_temp = new symbol('',0,'');
				for(var i = 0 ;i< length;i++){
					 obj = new symbol('',0,'');
					char.push(obj);
				}
				for(var i = 0; i < length; i++)
			    {    
			        for(var j=0; j< length; j++)
			        {
			            if (text[i]==char[j].str)
			            {
			                char[j].count++;
			                break;               
			            }
			            if (char[j].str=='')
			            {
			            	obj_str = new code('','');
							char[j].str=text[i];
							obj_str.str = text[i];
							string_arr.push(obj_str);
			                char[j].count=1;
			                count++;
			        		break;
			            }           
		        	}
		              
	    		} 
				bulbSort(char,count);
				consoleLog(char,count);
				temp = count-1;
				console.log(temp-1);
				for(var i = 0; i < count-1; i++){
					char = huffmanMethod(char,string_arr,temp);
					bulbSort(char,temp+1);
					temp--;
				}			document.getElementById('container').innerHTML = container;
				for(var i = 0; i< string_arr.length ;i++){
					console.log('code: '+string_arr[i].code+' char: '+string_arr[i].str);
				}
			}
			else if(!method){
			var container = '';
			var text1;
			//text = form.inputName.value;
			text = text.split('');
			length = text.length;
			for(var i = 0 ;i< length;i++){
				var obj_str = new symbolShennon('','',0);
				string_arr.push(obj_str);
			}
			for(var i = 0; i < length; i++)
		    {    
		        for(var j=0; j< length; j++)
		        {
		            if (text[i]==string_arr[j].str)
		            {
		                string_arr[j].count++;
		                break;               
		        
		            }
		            if (string_arr[j].str=='')
		            {
						string_arr[j].str=text[i];
		                string_arr[j].count=1;
		                count++;
		        		break;
		            }           
	        	}
	              
    		} 
    		for(var i = 0 ; i < string_arr.length; i++){
    			string_arr[i].count = string_arr[i].count/length;
    		}
			bulbSort(string_arr,count);
			Shannon_Fano(0,count-1);
			consoleLog(string_arr,count);
			
			}	
		}
	</script>
</head>
<body>
	<div class="container-fluid">
		<div class="row">
			<div class="col-md-3">
				<ul>
					<li onclick="activateHuffman()"><span>Хаффман</span></li>
					<li onclick="activateShennon_Fano()"><span>Шеннона-Фано</span></li>
				</ul>
			</div>
			<div class="col-md-9">
				<div class="row">
					<div class="col-md-5 col-sm-4">
						<form  name="test" method="post" id="form">
							 <input id="Input_text" type="text" size="30" name="inputName" placeholder="Input text">
							 <br>
							 <input  id="btn" class="click-btn" type="button" value="Click" onclick="checkForm(document.getElementById('form'))">
							 <input id="btn" type="button" value="Reload" onclick="location.reload(true);">
							 <input id="btn" type="button" value="decoder" onclick="decoder(string_arr,text)">
							  <input id="btn" type="button" value="encoder" onclick="encoder(string_arr,text)">
							 <ol id="demo"></ol>
						</form>
					</div>	
					 <div class="col-md-7 col-sm-4">
						<div class="download">	
							 <label class="file_upload">
						        <span class="button">Выбрать</span>
						        <mark>Файл не выбран</mark>
						        <input type="file" accept='text/plain' onchange='openFile(event)'><br>>
						    </label>
							<div>
							  <pre id="result"></pre>
							</div>
						</div>
					</div> 
				</div>
				<div class="container">
				  	<h2>Results</h2>
				  	<div id="container"></div>
				</div>			
			</div>
		</div>
	</div>
	
	
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.0/jquery.min.js"></script>
</body>
</html>
